src/Microsoft.TestPlatform.Utilities/CodeCoverageRunSettingsProcessor.cs

```diff
@@ -0,0 +1,515 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.VisualStudio.TestPlatform.Utilities
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Text;
+    using System.Xml;
+    using System.Xml.XPath;
+
+    using Microsoft.VisualStudio.TestPlatform.ObjectModel;
+
+    /// <summary>
+    /// Represents the run settings processor for code coverage data collectors.
+    /// </summary>
+    public class CodeCoverageRunSettingsProcessor
+    {
+        #region Type Members
+        /// <summary>
+        /// Represents the exclusion type for code coverage run settings.
+        /// </summary>
+        private class Exclusion
+        {
+            /// <summary>
+            /// Represents the <see cref="XPathNavigator"/> style path of the exclusion type.
+            /// </summary>
+            private string path;
+
+            /// <summary>
+            /// Gets the path components for the current exclusion type.
+            /// </summary>
+            public IEnumerable<string> PathComponents { get; }
+
+            /// <summary>
+            /// Gets the exclusion rules for the current exclusion type.
+            /// </summary>
+            public IDictionary<string, XmlNode> ExclusionRules { get; }
+
+            /// <summary>
+            /// Gets the actual exclusion type path generated from the individual path components.
+            /// </summary>
+            public string Path
+            {
+                get
+                {
+                    if (string.IsNullOrEmpty(this.path))
+                    {
+                        this.path = Exclusion.BuildPath(this.PathComponents);
+                    }
+
+                    return this.path;
+                }
+            }
+
+            /// <summary>
+            /// Constructs an <see cref="Exclusion"/> object.
+            /// </summary>
+            /// 
+            /// <param name="pathComponents">The path split in components.</param>
+            public Exclusion(IEnumerable<string> pathComponents)
+            {
+                this.path = string.Empty;
+                this.PathComponents = pathComponents;
+                this.ExclusionRules = new Dictionary<string, XmlNode>();
+            }
+
+            /// <summary>
+            /// Assembles a relative path from the path given as components.
+            /// </summary>
+            /// 
+            /// <returns>A relative path built from path components.</returns>
+            public static string BuildPath(IEnumerable<string> pathComponents)
+            {
+                var path = ".";
+
+                foreach (var component in pathComponents)
+                {
+                    path += "/" + component;
+                }
```

Comment:
This part can be improved by using StringBuilder.

Reply:
**TL;DR**: That's correct indeed and I did the change.

Long version:
Did it this way because the input is completely controller by us (even though the method is public the class in which it is defined is a private member of another class -- see [this](https://github.com/microsoft/vstest/pull/2431/files#diff-35b75052fa805261fcf83e696f1a45ccR23)). Moreover, the method is only called with arguments that aren't larger than 2 string components really. Felt it's more expressive this way than by using a StringBuilder at virtually no performance cost.